File: rust/tests/integration_test.rs
use aws_config::{BehaviorVersion, Region};
use aws_credential_types::{Credentials, provider::SharedCredentialsProvider};
use aws_lambdas_rust::{authenticate, signin, signup};
use aws_sdk_dynamodb::{types::{AttributeDefinition, KeySchemaElement, KeyType, ProvisionedThroughput, ScalarAttributeType}, Client};
use serde_json::{json, Value};
use testcontainers::clients;
use testcontainers_modules::localstack::LocalStack;

#[tokio::test]
async fn test_integration() {
    let docker = clients::Cli::default();
    let node = docker.run(LocalStack::default());
    let host_port = node.get_host_port_ipv4(4566);
    let endpoint_url = format!("http://127.0.0.1:{}", host_port);

    let credentials = Credentials::new("test", "test", None, None, "static");
    let credentials_provider = SharedCredentialsProvider::new(credentials);

    let config = aws_config::load_defaults(BehaviorVersion::latest())
        .await
        .into_builder()
        .region(Region::new("us-east-1"))
        .endpoint_url(&endpoint_url)
        .credentials_provider(credentials_provider)
        .build();

    let client = Client::new(&config);
    let table_name = "Users";
    let jwt_secret = "testsecret";

    client
        .create_table()
        .table_name(table_name)
        .key_schema(
            KeySchemaElement::builder()
                .attribute_name("email")
                .key_type(KeyType::Hash)
                .build()
                .unwrap(),
        )
        .attribute_definitions(
            AttributeDefinition::builder()
                .attribute_name("email")
                .attribute_type(ScalarAttributeType::S)
                .build()
                .unwrap(),
        )
        .provisioned_throughput(
            ProvisionedThroughput::builder()
                .read_capacity_units(5)
                .write_capacity_units(5)
                .build()
                .unwrap(),
        )
        .send()
        .await
        .expect("Failed to create table");

    let signup_body = json!({
        "email": "test@example.com",
        "password": "password123",
        "name": "Test User"
    }).to_string();

    let resp = signup(&signup_body, &client, table_name).await.unwrap();
    let status = resp["statusCode"].as_i64().unwrap();
    assert_eq!(status, 201);

    let resp = signup(&signup_body, &client, table_name).await.unwrap();
    let status = resp["statusCode"].as_i64().unwrap();
    assert_eq!(status, 409);

    let signin_body = json!({
        "email": "test@example.com",
        "password": "password123"
    }).to_string();

    let resp = signin(&signin_body, &client, table_name, jwt_secret).await.unwrap();
    let status = resp["statusCode"].as_i64().unwrap();
    assert_eq!(status, 200);

    let body_str = resp["body"].as_str().unwrap();
    let body_json: Value = serde_json::from_str(body_str).unwrap();
    let token = body_json["token"].as_str().unwrap();

    let headers = json!({
        "Authorization": format!("Bearer {}", token)
    });

    let resp = authenticate(&headers, jwt_secret).unwrap();
    let status = resp["statusCode"].as_i64().unwrap();
    assert_eq!(status, 200);

    let headers = json!({
        "Authorization": "Bearer invalidtoken"
    });
    let resp = authenticate(&headers, jwt_secret).unwrap();
    let status = resp["statusCode"].as_i64().unwrap();
    assert_eq!(status, 401);
}

File: nodejs/integration.test.mjs
import { GenericContainer, Wait } from "testcontainers";
import { DynamoDBClient, CreateTableCommand } from "@aws-sdk/client-dynamodb";
import { handler } from "./index.mjs";

describe("Lambda Integration Tests", () => {
  let container;
  let dynamoClient;
  const tableName = "Users";
  const jwtSecret = "testsecret";

  beforeAll(async () => {
    // Start LocalStack container
    container = await new GenericContainer("localstack/localstack:3.2.0")
      .withExposedPorts(4566)
      .withEnvironment({ SERVICES: "dynamodb" })
      .withWaitStrategy(Wait.forLogMessage("Ready."))
      .start();

    const endpoint = `http://${container.getHost()}:${container.getMappedPort(4566)}`;

    // Configure Environment Variables
    process.env.TABLE_NAME = tableName;
    process.env.JWT_SECRET = jwtSecret;
    // Override AWS SDK endpoint via env var is tricky without re-importing,
    // but since we instantiate client in global scope of index.mjs, we might need to mock or rely on AWS_ENDPOINT_URL if SDK supports it.
    // AWS SDK v3 supports AWS_ENDPOINT_URL env var.
    process.env.AWS_ENDPOINT_URL = endpoint;
    process.env.AWS_REGION = "us-east-1";
    process.env.AWS_ACCESS_KEY_ID = "test";
    process.env.AWS_SECRET_ACCESS_KEY = "test";

    // Initialize DynamoDB Client for setup
    dynamoClient = new DynamoDBClient({
      endpoint,
      region: "us-east-1",
      credentials: { accessKeyId: "test", secretAccessKey: "test" },
    });

    // Create Table
    await dynamoClient.send(
      new CreateTableCommand({
        TableName: tableName,
        KeySchema: [{ AttributeName: "email", KeyType: "HASH" }],
        AttributeDefinitions: [{ AttributeName: "email", AttributeType: "S" }],
        ProvisionedThroughput: { ReadCapacityUnits: 5, WriteCapacityUnits: 5 },
      })
    );
  }, 60000);

  afterAll(async () => {
    if (container) await container.stop();
  });

  test("Signup - Success", async () => {
    const event = {
      httpMethod: "POST",
      path: "/signup",
      body: JSON.stringify({
        email: "test@example.com",
        password: "password123",
        name: "Test User",
      }),
    };

    const response = await handler(event);
    expect(response.statusCode).toBe(201);
    const body = JSON.parse(response.body);
    expect(body.message).toBe("User created successfully.");
  });

  test("Signup - Duplicate", async () => {
    const event = {
      httpMethod: "POST",
      path: "/signup",
      body: JSON.stringify({
        email: "test@example.com",
        password: "password123",
        name: "Test User",
      }),
    };

    const response = await handler(event);
    expect(response.statusCode).toBe(409);
  });

  test("Signin - Success", async () => {
    const event = {
      httpMethod: "POST",
      path: "/signin",
      body: JSON.stringify({
        email: "test@example.com",
        password: "password123",
      }),
    };

    const response = await handler(event);
    expect(response.statusCode).toBe(200);
    const body = JSON.parse(response.body);
    expect(body.token).toBeDefined();
  });

  test("Signin - Invalid Credentials", async () => {
    const event = {
      httpMethod: "POST",
      path: "/signin",
      body: JSON.stringify({
        email: "test@example.com",
        password: "wrongpassword",
      }),
    };

    const response = await handler(event);
    expect(response.statusCode).toBe(401);
  });

  test("Authentication - Success", async () => {
    // Get token first
    const signinEvent = {
      httpMethod: "POST",
      path: "/signin",
      body: JSON.stringify({
        email: "test@example.com",
        password: "password123",
      }),
    };
    const signinResponse = await handler(signinEvent);
    const token = JSON.parse(signinResponse.body).token;

    const event = {
      httpMethod: "POST",
      path: "/authentication",
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };

    const response = await handler(event);
    expect(response.statusCode).toBe(200);
  });

  test("Authentication - Invalid Token", async () => {
    const event = {
      httpMethod: "POST",
      path: "/authentication",
      headers: {
        Authorization: "Bearer invalidtoken",
      },
    };

    const response = await handler(event);
    expect(response.statusCode).toBe(401);
  });
});
